runtime/vm/vmprops.c

```diff
@@ -1433,3 +1445,648 @@ getMUtf8String(J9JavaVM *vm, const char *userString, UDATA stringLength)
 	}
 	return result;
 }
+
+/**
+ * @return JNI_FALSE on success, JNI_TRUE on error
+ */
+static jboolean
+propertyListAddString(JNIEnv *env, jarray array, jint arrayIndex, const char *value)
+{
+	/* String must be well-formed modified UTF-8 */
+	jobject str = (*env)->NewStringUTF(env, value);
+	if (NULL != str) {
+		(*env)->SetObjectArrayElement(env, array, arrayIndex, str);
+	}
+	/* NewStringUTF does not throw an exception, other than OutOfMemory */
+	return (*env)->ExceptionCheck(env);
+}
+
+typedef struct {
+	jboolean errorOccurred;
+	jobject args;
+	jint nCommandLineDefines;
+	JNIEnv *env;
+	const char **defaultValues;
+	int defaultCount;
+	jclass stringClass;
+} CreateSystemPropertiesData;
+
+static void JNICALL
+systemPropertyIterator(char *key, char *value, void *userData)
+{
+	CreateSystemPropertiesData *iteratorData = userData;
+	jobject args = iteratorData->args;
+	JNIEnv *env = iteratorData->env;
+	const char **defaultValues = iteratorData->defaultValues;
+	int defaultCount = iteratorData->defaultCount;
+	jint i = 0;
+
+	/* CMVC 95717: if an error has already occurred get out of here */
+	if (iteratorData->errorOccurred) {
+		return;
+	}
+
+	if (0 == strcmp("com.ibm.oti.shared.enabled", key)) {
+		/* JAZZ103 85641: Prevent com.ibm.oti.shared.enabled from being overwritten by a command line option */
+		return;
+	}
+
+#if JAVA_SPEC_VERSION >= 21
+	if (0 == strcmp("java.compiler", key)) {
+		PORT_ACCESS_FROM_ENV(env);
+		if ((0 == strcmp("jitc", value)) || (0 == strcmp(J9_JIT_DLL_NAME, value))) {
+			j9nls_printf(PORTLIB, J9NLS_WARNING, J9NLS_VM_JAVA_COMPILER_WARNING_XJIT);
+		} else {
+			j9nls_printf(PORTLIB, J9NLS_WARNING, J9NLS_VM_JAVA_COMPILER_WARNING_XINT);
+		}
+		return;
+	}
+#endif /* JAVA_SPEC_VERSION >= 21 */
+
+	/* check for overridden system properties, use linear scan for now */
+	for (i = 0; i < defaultCount; i += 2) {
+		if ((NULL != defaultValues[i]) && (0 == strcmp(key, defaultValues[i]))) {
+			defaultValues[i] = NULL;
+			defaultValues[i + 1] = NULL;
+			break;
+		}
+	}
+
+	/* First do the key */
+	if (propertyListAddString(env, args, iteratorData->nCommandLineDefines++, key)) {
+		iteratorData->errorOccurred = 1;
```

Comment:
Only `JNI_TRUE` or `JNI_FALSE` should be written to `errorOccurred`.

Reply:
Updated.